mongostore.go

```diff
@@ -0,0 +1,264 @@
+	SenderSubId      string `bson:"sender_sub_id"`
+	SenderLocId      string `bson:"sender_loc_id"`
+	TargetCompId     string `bson:"target_comp_id"`
+	TargetSubId      string `bson:"target_sub_id"`
+	TargetLocId      string `bson:"target_loc_id"`
+}
+
+// Reset deletes the store records and sets the seqnums back to 1
+func (store *mongoStore) Reset() error {
+	msgFilter := generateMessageFilter(&store.sessionID)
+	_, err := store.db.DB(store.mongoDatabase).C(store.messagesCollection).RemoveAll(msgFilter)
+
+	if err != nil {
+		return err
+	}
+
+	if err = store.cache.Reset(); err != nil {
+		return err
+	}
+
+	sessionUpdate := generateMessageFilter(&store.sessionID)
+	sessionUpdate.CreationTime = store.cache.CreationTime()
+	sessionUpdate.IncomingSeqNum = store.cache.NextTargetMsgSeqNum()
+	sessionUpdate.OutgoingSeqNum = store.cache.NextSenderMsgSeqNum()
+	err = store.db.DB(store.mongoDatabase).C(store.sessionsCollection).Update(msgFilter, sessionUpdate)
+
+	return err
+}
+
+// Refresh reloads the store from the database
+func (store *mongoStore) Refresh() error {
+	if err := store.cache.Reset(); err != nil {
+		return err
+	}
+	return store.populateCache()
+}
+
+func (store *mongoStore) populateCache() (err error) {
+	msgFilter := generateMessageFilter(&store.sessionID)
+	query := store.db.DB(store.mongoDatabase).C(store.sessionsCollection).Find(msgFilter)
+
+	if cnt, err := query.Count(); err == nil && cnt > 0 {
+		// session record found, load it
+		sessionData := &MongoQuickFixEntryData{}
+		err = query.One(&sessionData)
+		if err == nil {
+			store.cache.creationTime = sessionData.CreationTime
+			store.cache.SetNextTargetMsgSeqNum(sessionData.IncomingSeqNum)
+			store.cache.SetNextSenderMsgSeqNum(sessionData.OutgoingSeqNum)
+		}
+	} else if err == nil && cnt == 0 {
+		// session record not found, create it
+		msgFilter.CreationTime = store.cache.creationTime
+		msgFilter.IncomingSeqNum = store.cache.NextTargetMsgSeqNum()
+		msgFilter.OutgoingSeqNum = store.cache.NextSenderMsgSeqNum()
+		err = store.db.DB(store.mongoDatabase).C(store.sessionsCollection).Insert(msgFilter)
+	}
+	return
+}
+
+// NextSenderMsgSeqNum returns the next MsgSeqNum that will be sent
+func (store *mongoStore) NextSenderMsgSeqNum() int {
+	return store.cache.NextSenderMsgSeqNum()
+}
+
+// NextTargetMsgSeqNum returns the next MsgSeqNum that should be received
+func (store *mongoStore) NextTargetMsgSeqNum() int {
+	return store.cache.NextTargetMsgSeqNum()
+}
+
+// SetNextSenderMsgSeqNum sets the next MsgSeqNum that will be sent
+func (store *mongoStore) SetNextSenderMsgSeqNum(next int) error {
+	msgFilter := generateMessageFilter(&store.sessionID)
+	sessionUpdate := generateMessageFilter(&store.sessionID)
+	sessionUpdate.IncomingSeqNum = store.cache.NextTargetMsgSeqNum()
+	sessionUpdate.OutgoingSeqNum = next
+	sessionUpdate.CreationTime = store.cache.CreationTime()
+	err := store.db.DB(store.mongoDatabase).C(store.sessionsCollection).Update(msgFilter, sessionUpdate)
```

Comment:
can combine this with next line

Reply:
[https://github.com/quickfixgo/quickfix/pull/343/commits/2ae607bc4bb0cbb95140fe6501dba4c364cc5003](
https://github.com/quickfixgo/quickfix/pull/343/commits/2ae607bc4bb0cbb95140fe6501dba4c364cc5003)